### PR TITLE
feat(cli): support spfx teamsfx preview

### PR DESCRIPTION
**Pull request by jiayi-ruan**
Add support for spfx both local/remote preview.

related task:
https://msazure.visualstudio.com/Microsoft%20Teams%20Extensibility/_sprints/taskboard/Teams%20Extensibility%20E2E%20Team/Microsoft%20Teams%20Extensibility/CY21-7.2?workitem=10386273
https://msazure.visualstudio.com/Microsoft%20Teams%20Extensibility/_sprints/taskboard/Teams%20Extensibility%20E2E%20Team/Microsoft%20Teams%20Extensibility/CY21-7.2?workitem=10386275

### CODE REVIEW

**Review by Alive-Fish — APPROVED** [review]

**Review by chagong — APPROVED** [review]

File: packages/cli/src/cmds/preview/preview.ts [hunk 1]
```diff
@@ -316,9 +501,12 @@ export default class Preview extends YargsCommand {
     const includeSpfx = activeResourcePlugins.some(
       (pluginName) => pluginName === constants.spfxPluginName
     );
-    // TODO: remove when SPFx preview is ready
     if (includeSpfx) {
-      return err(errors.SPFxNotSupported());
+      if (!this.sharepointSiteUrl) {
+        return err(errors.NoUrlForSPFxRemotePreview());
+      }
+      const spfxRoot = path.join(workspaceFolder, constants.spfxFolderName);
```
Comment (kuojianlu) [inline packages/cli/src/cmds/preview/preview.ts:508]: Check `spfxRoot`.

File: packages/cli/src/cmds/preview/preview.ts [hunk 2]
```diff
@@ -289,6 +309,171 @@ export default class Preview extends YargsCommand {
...
+      constants.gulpServeTitle,
+      true,
+      constants.nodeCommand,
+      [`${spfxRoot}/node_modules/gulp/bin/gulp.js`, "serve", "--nobrowser", "--no-color"],
+      {
+        shell: false,
+        cwd: spfxRoot,
+      }
+    );
+
```
Comment (kuojianlu) [inline packages/cli/src/cmds/preview/preview.ts:386]: Add `this.backgroundTasks.push(gulpServeTask);` here.

**Comment by kuojianlu** [comment]
Better to attach images to show the running result.

File: packages/cli/src/cmds/preview/preview.ts [hunk 3]
```diff
@@ -93,6 +99,19 @@ export default class Preview extends YargsCommand {
...
+          let spSite = args["sharepoint-site"] as string;
+          if (!spSite.startsWith("https")) {
+            spSite = `https://${spSite}`;
+          }
+          const spWorkbenchHttpsUrl = new URL("_layouts/workbench.aspx", spSite);
+          this.sharepointSiteUrl = spWorkbenchHttpsUrl.toString();
+        } catch (error) {
+          throw errors.InvalidSharePointSiteURL(error);
+        }
+      }
```
[the remark this answers is not in the document]
Comment (jiayi-ruan) [reply packages/cli/src/cmds/preview/preview.ts:126]: resolved. thx.

File: packages/cli/src/cmds/preview/preview.ts (hunk 2 above)
Comment (jiayi-ruan) [reply packages/cli/src/cmds/preview/preview.ts:387]: resolved. thx.

File: packages/cli/src/cmds/preview/preview.ts (hunk 1 above)
Comment (jiayi-ruan) [reply packages/cli/src/cmds/preview/preview.ts:512]: resolved. thx.

File: packages/cli/src/cmds/preview/preview.ts [hunk 4]
```diff
@@ -166,9 +185,10 @@ export default class Preview extends YargsCommand {
     const includeSpfx = activeResourcePlugins.some(
       (pluginName) => pluginName === constants.spfxPluginName
     );
-    // TODO: remove when SPFx preview is ready
+
     if (includeSpfx) {
-      return err(errors.SPFxNotSupported());
+      const spfxRoot = path.join(workspaceFolder, constants.spfxFolderName);
+      return this.spfxPreview(spfxRoot, browser, "https://localhost:5432/workbench");
```
[the remark this answers is not in the document]
Comment (jiayi-ruan) [reply packages/cli/src/cmds/preview/preview.ts:191]: resolved. thx.

**Review by kuojianlu — APPROVED** [review]

**Pull request merged**